inform7/Internal/Inter/Architecture32Kit/Sections/Glk.i6t

```diff
@@ -4,1001 +4,230 @@ Glk is a fully-featured input-output model, designed by Andrew Plotkin for use
-  @glk 360 _vararg_count 0;
-  return 0;
+[ IDENTIFY_FILEREFS_R;
+	switch (current_glk_object_rock) {
+		GG_SCRIPTFREF_ROCK:
+			gg_scriptfref = current_glk_object_reference;
+	}
+	rfalse;
 ];
 
-[ glk_time_to_date_local _vararg_count;
-  ! glk_time_to_date_local(&{int, uint, int}, &{int, int, int, int, int, int, int, int})
-  @glk 361 _vararg_count 0;
-  return 0;
+[ IDENTIFY_SCHANNELS_R;
+	switch (current_glk_object_rock) {
+		GG_FOREGROUNDCHAN_ROCK:
+			gg_foregroundchan = current_glk_object_reference;
+		GG_BACKGROUNDCHAN_ROCK:
+			gg_backgroundchan = current_glk_object_reference;
+	}
+	rfalse;
 ];
 
-[ glk_simple_time_to_date_utc _vararg_count;
-  ! glk_simple_time_to_date_utc(int, uint, &{int, int, int, int, int, int, int, int})
-  @glk 362 _vararg_count 0;
-  return 0;
+[ STOP_SCHANNELS_R;
+	if (glk_gestalt(gestalt_Sound, 0)) {
+		if (gg_foregroundchan ~= 0) {
+			glk_schannel_stop(gg_foregroundchan);
+		}
+		if (gg_backgroundchan ~= 0) {
+			glk_schannel_stop(gg_backgroundchan);
+		}
+	}
+	rfalse;
 ];
 
-[ glk_simple_time_to_date_local _vararg_count;
-  ! glk_simple_time_to_date_local(int, uint, &{int, int, int, int, int, int, int, int})
-  @glk 363 _vararg_count 0;
-  return 0;
-];
+@h Glk windows.
+Functions for minimal Glk windows support.
 
-[ glk_date_to_time_utc _vararg_count;
-  ! glk_date_to_time_utc(&{int, int, int, int, int, int, int, int}, &{int, uint, int})
-  @glk 364 _vararg_count 0;
-  return 0;
+=
+[ FindGlkWindowFromRefNum ref win;
+	if (ref ~= 0) {
+		for (win = K_Glk_Window_First: win: win = win.K_Glk_Window_Next) {
+			if (win.glk_ref == ref) {
+				return win;
+			}
+		}
+	}
+	return nothing;
 ];
 
-[ glk_date_to_time_local _vararg_count;
-  ! glk_date_to_time_local(&{int, int, int, int, int, int, int, int}, &{int, uint, int})
-  @glk 365 _vararg_count 0;
-  return 0;
-];
+@h Glk events.
+To handle events we intercept the |glk_select| function. This allows us to handle
+events early and consistently.
 
-[ glk_date_to_simple_time_utc _vararg_count ret;
-  ! glk_date_to_simple_time_utc(&{int, int, int, int, int, int, int, int}, uint) => int
-  @glk 366 _vararg_count ret;
-  return ret;
-];
+=
+Global Glk_Event_Struct_type;
+Global Glk_Event_Struct_win;
+Global Glk_Event_Struct_val1;
+Global Glk_Event_Struct_val2;
+
+[ glk_select event_struct;
+	! Call the real glk_select
+	@push event_struct;
+	@glk 192 1 0;
+
+	! Clear the unknown window's ref so that we don't immediately match it
+	Unknown_Glk_Window.glk_ref = 0;
+
+	! Copy values to our variables
+	Glk_Event_Struct_type = event_struct-->0;
+	Glk_Event_Struct_win = FindGlkWindowFromRefNum(event_struct-->1);
+	Glk_Event_Struct_val1 = event_struct-->2;
+	Glk_Event_Struct_val2 = event_struct-->3;
+
+	! If no window was found, then use the unknown window
+	if (Glk_Event_Struct_win == nothing && event_struct-->1) {
+		Glk_Event_Struct_win = Unknown_Glk_Window;
+		Unknown_Glk_Window.glk_ref = event_struct-->1;
```

Comment:
While most people wanting to add extra windows would probably do it with Flexible Windows, there's still the possibility of a window being defined through I6 code that doesn't have an I7 glk window object. This function converts the window reference ID to a window object for the glk event handling rules, and then converts it back to a reference ID at the end. If we couldn't find a window object and just passed in `nothing`, then we would end up wiping the reference ID at the end. So the solution I've come up with is to use a dummy object, the unknown window. If we can't find another window object we use this one, updating its reference ID and rock. Then at the end we can pull the window reference out of it again, allowing us to preserve the event struct if the author didn't manually change it.